collector/jsz.go

```diff
@@ -0,0 +1,327 @@
+// Copyright 2021 The NATS Authors
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Package collector has various collector utilities and implementations.
+package collector
+
+import (
+	"net/http"
+	"sync"
+	"time"
+
+	nats "github.com/nats-io/nats-server/v2/server"
+	"github.com/prometheus/client_golang/prometheus"
+)
+
+const (
+	jszSuffix = "/jsz?consumers=true"
+)
+
+type jszCollector struct {
+	sync.Mutex
+	httpClient *http.Client
+	servers    []*CollectedServer
+
+	// JetStream server stats
+	streams    *prometheus.Desc
+	consumers  *prometheus.Desc
+	messages   *prometheus.Desc
+	bytes      *prometheus.Desc
+	maxMemory  *prometheus.Desc
+	maxStorage *prometheus.Desc
+
+	// Stream stats
+	streamMessages      *prometheus.Desc
+	streamBytes         *prometheus.Desc
+	streamLastSeq       *prometheus.Desc
+	streamConsumerCount *prometheus.Desc
+
+	// Consumer stats
+	consumerDeliveredConsumerSeq *prometheus.Desc
+	consumerDeliveredStreamSeq   *prometheus.Desc
+	consumerNumAckPending        *prometheus.Desc
+	consumerNumRedelivered       *prometheus.Desc
+	consumerNumWaiting           *prometheus.Desc
+	consumerNumPending           *prometheus.Desc
+}
+
+func isJszEndpoint(system string) bool {
+	return system == JetStreamSystem
+}
+
+func newJszCollector(system string, servers []*CollectedServer) prometheus.Collector {
+	serverLabels := []string{"server_id", "server_name", "cluster", "domain", "meta_leader", "is_meta_leader"}
+
+	var streamLabels []string
+	streamLabels = append(streamLabels, serverLabels...)
+	streamLabels = append(streamLabels, "stream_name")
+	streamLabels = append(streamLabels, "stream_leader")
+	streamLabels = append(streamLabels, "is_stream_leader")
+
+	var consumerLabels []string
+	consumerLabels = append(consumerLabels, streamLabels...)
+	consumerLabels = append(consumerLabels, "consumer_name")
```

Comment:
This will probably lead to massive explosions in prometheus storage, there will simply be too many consumer names, to a lesser extend but still a concern also per stream values

https://www.robustperception.io/cardinality-is-key

Reply:
I'll make this opt-in, think some of this would be needed to enable horizontal autoscaler use cases though maybe prometheus not the best tool for that
```
jetstream_consumer_num_pending{is_consumer_leader="true",consumer_name="bar"}
jetstream_consumer_num_ack_pending{is_consumer_leader="true",consumer_name="bar"}
```